### PR TITLE
feat(Nexus HA): determination of nexus destination node on failover

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**
This adds the node determination module in case of implicit failover, scheduling decision is made based on the following information:
- topology restrictions for volume.
- distribution of volumes across nodes (to provide equal volume distribution across nodes).

Signed-off-by: Abhinandan Purkait <abhinandan.purkait@mayadata.io>

### CODE REVIEW

File: control-plane/agents/core/src/core/scheduling/mod.rs [hunk 1]
```diff
@@ -45,8 +45,8 @@ pub(crate) trait ResourceFilter: Sized {
 /// Filter nodes used for replica creation
 pub(crate) struct NodeFilters {}
 impl NodeFilters {
-    /// Should only attempt to use online nodes
-    pub(crate) fn online(_request: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
+    /// Should only attempt to use online nodes for pools
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/scheduling/mod.rs]: nit: end all doc comments with a period `.`

**Comment by Abhinandan-Purkait** [comment]
bors try

File: control-plane/agents/core/src/core/scheduling/mod.rs [hunk 2]
```diff
@@ -291,3 +320,24 @@ impl AddReplicaSorters {
         }
     }
 }
+
+/// sort nodes to pick the best choice for nexus target
+pub(crate) struct NodeSorters {}
+impl NodeSorters {
+    /// sort nodes by the number of active nexus present per node,
+    /// the lesser the number of active nexus on a node, the more would be its selection priority
+    pub(crate) fn sort_by_active_targets_present(a: &NodeItem, b: &NodeItem) -> std::cmp::Ordering {
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/scheduling/mod.rs]: rename suggestion: `number_targets`

File: control-plane/agents/core/src/core/scheduling/mod.rs [hunk 3]
```diff
@@ -291,3 +320,24 @@ impl AddReplicaSorters {
...
+        a.node_wrapper
+            .nexus_states()
+            .into_iter()
+            .filter(|n| n.nexus.status == NexusStatus::Online)
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/scheduling/mod.rs]: why does the status matter?

File: control-plane/agents/core/src/core/scheduling/resources/mod.rs [hunk 4]
```diff
@@ -194,3 +194,20 @@ impl ChildItem {
         &self.pool_state
     }
 }
+
+/// Individual Node candidate which is a wrapper over nodewrapper used for filtering
+#[derive(Clone)]
+pub(crate) struct NodeItem {
+    pub(crate) node_wrapper: NodeWrapper,
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/scheduling/resources/mod.rs]: you could expose this via a function instead and leave it private

File: control-plane/agents/core/src/nexus/scheduling.rs [hunk 5]
```diff
@@ -43,3 +47,24 @@ pub(crate) async fn get_healthy_nexus_children(
 
     Ok(children)
 }
+
+/// Return the suitable node target to publish the volume for nexus placement on
+/// volume publish or nexus failover
+pub(crate) async fn get_nexus_target_node_candidate(
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/nexus/scheduling.rs]: suggestion: `get_target_node_candidate`

File: control-plane/agents/core/src/volume/specs.rs [hunk 6]
```diff
@@ -1504,16 +1505,11 @@ async fn get_volume_target_node(
...
-            let nodes = registry.get_node_wrappers().await;
-            for locked_node in nodes {
-                let node = locked_node.read().await;
-                // todo: use other metrics in order to make the "best" choice
-                if node.is_online() {
-                    return Ok(node.id().clone());
-                }
-            }
-            Err(SvcError::NoNodes {})
+            // incase there is no target node specified, let the control-plane scheduling logic
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/volume/specs.rs]: nit: `in case`

File: control-plane/agents/core/src/core/scheduling/mod.rs [hunk 7]
```diff
@@ -68,6 +68,35 @@ impl NodeFilters {
...
             .any(|node_spec| node_spec.id() == &item.pool.node)
     }
+
+    /// Should only attempt to use online nodes
+    pub(crate) fn online(_request: &GetSuitableNodesContext, item: &NodeItem) -> bool {
+        item.node_wrapper.is_online()
+    }
+
+    /// Should abide by the topology constraints
+    pub(crate) fn topology(request: &GetSuitableNodesContext, item: &NodeItem) -> bool {
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/scheduling/mod.rs]: We shouldn't add this, the topology is for data replicas and not nexuses. And until we decide what to do with it we shouldn't be adding more code using it.

File: control-plane/agents/core/src/core/scheduling/mod.rs [hunk 8]
```diff
@@ -45,8 +45,8 @@ pub(crate) trait ResourceFilter: Sized {
 /// Filter nodes used for replica creation
 pub(crate) struct NodeFilters {}
 impl NodeFilters {
-    /// Should only attempt to use online nodes
-    pub(crate) fn online(_request: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
+    /// Should only attempt to use online nodes for pools
+    pub(crate) fn online_for_pool(_request: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/core/scheduling/mod.rs]: Sorry this is my fault, this should have been trait based, we'll address this later.

Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/core/scheduling/mod.rs]: Sure.

File: control-plane/agents/core/src/core/scheduling/mod.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/core/scheduling/mod.rs]: Done.

File: control-plane/agents/core/src/core/scheduling/mod.rs (hunk 2 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/core/scheduling/mod.rs]: `sort_by_active_targets_present` -> `number_targets`

File: control-plane/agents/core/src/core/scheduling/mod.rs (hunk 3 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/core/scheduling/mod.rs]: Removing it for now.

File: control-plane/agents/core/src/core/scheduling/resources/mod.rs (hunk 4 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/core/scheduling/resources/mod.rs]: Done.

File: control-plane/agents/core/src/nexus/scheduling.rs (hunk 5 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/nexus/scheduling.rs]: `get_nexus_target_node_candidate` -> `get_target_node_candidate`

File: control-plane/agents/core/src/core/scheduling/mod.rs (hunk 7 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/core/scheduling/mod.rs]: Removing it for now as we are not honoring any topology. Need to revisit this once we have the changes.
cc @mynktl @mtzaurus

**Review by mynktl — APPROVED** [review]

File: control-plane/agents/core/src/volume/specs.rs [hunk 9]
```diff
@@ -1112,16 +1114,12 @@ pub(crate) async fn get_volume_target_node(
...
-                    return Ok(node.id().clone());
-                }
-            }
-            Err(SvcError::NoNodes {})
+            // in case there is no target node specified, let the control-plane scheduling logic
+            // determine a suitable node for the same.
+            let volume_spec = registry.specs().get_volume(&status.uuid)?;
+            let candidate = get_target_node_candidate(&volume_spec, registry).await?;
+            info!(node.id=%candidate.id(), "node selected for volume publish by control-plane");
+            return Ok(candidate.id().clone());
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/volume/specs.rs]:
```suggestion
            Ok(candidate.id().clone())
```

File: control-plane/agents/core/src/volume/specs.rs [hunk 10]
```diff
@@ -1112,16 +1114,12 @@ pub(crate) async fn get_volume_target_node(
...
-                if node.is_online() {
-                    return Ok(node.id().clone());
-                }
-            }
-            Err(SvcError::NoNodes {})
+            // in case there is no target node specified, let the control-plane scheduling logic
+            // determine a suitable node for the same.
+            let volume_spec = registry.specs().get_volume(&status.uuid)?;
+            let candidate = get_target_node_candidate(&volume_spec, registry).await?;
+            info!(node.id=%candidate.id(), "node selected for volume publish by control-plane");
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/volume/specs.rs]: this should be debug

**Review by tiagolobocastro — APPROVED** [review]

File: control-plane/agents/core/src/volume/specs.rs (hunk 9 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/volume/specs.rs]: Done.

File: control-plane/agents/core/src/volume/specs.rs (hunk 10 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/core/src/volume/specs.rs]: Done

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**